File: src/main/java/mygame/editor/model/PolygonModel.java
package mygame.editor.model;

import javafx.geometry.Point2D;
import javafx.scene.paint.*;
import javafx.scene.shape.*;
import javafx.scene.shape.Polygon;
import mygame.Constants;

import java.util.*;

/**
 * Created by oleh on 3/21/17.
 */
public class PolygonModel extends AbstractModel {
    Polygon polygon;
    List<Point> points;


    public PolygonModel(Point2D startPoint){
        super(Type.POLYGON);
        points = new ArrayList<>();
        Point point = new Point(startPoint);
        points.add(point);
        polygon = new Polygon();
        polygon.setFill(Constants.ORANGE.deriveColor(1, 1, 1, .25));
        polygon.setStroke(Constants.ORANGE);
        getChildren().add(polygon);
        Circle handler = getHandler(startPoint);
        handler.setUserData(point);
        addControl(handler);
        polygon.getPoints().addAll(startPoint.getX(), startPoint.getY());


    }

    public void addPoint(Point2D newPoint){
        Point point = new Point(newPoint);
        Circle handler = getHandler(newPoint);
        handler.setUserData(point);
        addControl(handler);

        points.add(point);

       updatePolygon();
    }




    public void updatePolygon(){
        polygon.getPoints().clear();
        if(points.size() >2)
        for (Point point : points) {
            polygon.getPoints().add(point.getX());
            polygon.getPoints().add(point.getY());
        }
    }

    public void updateLastPoint(Point2D newLocation){

    }
}

File: src/main/java/mygame/editor/model/PolygonModel.kt
package mygame.editor.model

import javafx.scene.shape.Polygon
import mygame.Constants
import mygame.editor.model.CustonAffine

class PolygonModel constructor(startPoint: Point) : AbstractModel(Type.POLYGON) {
    var polygon: Polygon

    init {

        appendPoint(startPoint)
        polygon = Polygon()
        polygon.fill = Constants.ORANGE.deriveColor(1.0, 1.0, 1.0, .25)
        polygon.stroke = Constants.ORANGE
        children.add(polygon)



    }

    fun addPoint(newPoint:Point) {
        points.add(newPoint)
        update()
    }




    fun updateLastPoint(newLocation: Point) {

    }

    override fun transform(trans: CustonAffine) {
        polygon.points.clear()
        if (points.size > 2)
            for (point in points) {
                val p  = trans.transformPoint(point)
                polygon.points.add(p.x)
                polygon.points.add(p.y)
            }
    }

    override fun update(vararg points: Point) {

    }
}
